### PR TITLE
Implement Prysm-Specific HashTreeRootState

### PR DESCRIPTION
**Pull request by rauljordan**
No tracking issue.

---

# Description

**Write why you are making the changes in this pull request**

This PR implements a highly customized, Prysm-specific version of `HashTreeRoot` for the beacon state without usage of the `reflect` or `unsafe` package.

```go
BenchmarkHashTreeRoot_Old-4           14         129413837 ns/op        30069618 B/op     469708 allocs/op
BenchmarkHashTreeRoot_New-4           24          49718441 ns/op        16446592 B/op     160480 allocs/op
```
This version is approximately a 3x improvement across the board when tree hashing a mainnet beacon state. It currently passes latest spectests.

### Feature Flag

The new feature flag for this is `--enable-custom-state-ssz`

### CODE REVIEW

**Comment by prestonvanloon** [comment]
LGTM, but test failures

File: shared/stateutil/state_root_test.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import (
+	"testing"
+
+	"github.com/prysmaticlabs/go-ssz"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/interop"
+	"github.com/prysmaticlabs/prysm/shared/stateutil"
+)
+
+func TestHashTreeRootEquality(t *testing.T) {
```
Comment (0xKiwi) [inline shared/stateutil/state_root_test.go]: Would it be possible to test everything else? Benchmarks aren't needed but maybe splitting it up would make the tests cleaner too.

**Comment by 0xKiwi** [comment]
Any updates on this? @rauljordan

**Comment by rauljordan** [comment]
Hey Kiwi yep big priority for me after some API fixes

File: shared/stateutil/state_root.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+	if checkpoint != nil {
+		epochBuf := make([]byte, 8)
+		binary.LittleEndian.PutUint64(epochBuf, checkpoint.Epoch)
+		epochRoot := bytesutil.ToBytes32(epochBuf)
+		fieldRoots[0] = epochRoot[:]
+		fieldRoots[1] = checkpoint.Root
+	}
+	root, err := bitwiseMerkleize(fieldRoots, uint64(len(fieldRoots)), uint64(len(fieldRoots)))
+	if err != nil {
+		return [32]byte{}, nil
```
Comment (0xKiwi) [inline shared/stateutil/state_root.go]: Can you return the err here?

Comment (0xKiwi) [inline shared/stateutil/state_root.go]: Can you return the err here?

File: shared/stateutil/blocks.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	headerSlotBuf := make([]byte, 8)
+	binary.LittleEndian.PutUint64(headerSlotBuf, header.Slot)
+	headerSlotRoot := bytesutil.ToBytes32(headerSlotBuf)
+	fieldRoots[0] = headerSlotRoot[:]
+	fieldRoots[1] = header.ParentRoot
+	fieldRoots[2] = header.StateRoot
+	fieldRoots[3] = header.BodyRoot
+	signatureChunks, err := pack([][]byte{header.Signature})
+	if err != nil {
+		return [32]byte{}, nil
```
Comment (0xKiwi) [inline shared/stateutil/blocks.go]: same here

File: shared/stateutil/blocks.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+	}
+	sigRoot, err := bitwiseMerkleize(signatureChunks, uint64(len(signatureChunks)), uint64(len(signatureChunks)))
+	if err != nil {
+		return [32]byte{}, nil
+	}
+	fieldRoots[4] = sigRoot[:]
+	root, err := bitwiseMerkleize(fieldRoots, uint64(len(fieldRoots)), uint64(len(fieldRoots)))
+	if err != nil {
+		return [32]byte{}, nil
+	}
```
Comment (0xKiwi) [inline shared/stateutil/blocks.go]: And these two please

File: shared/stateutil/blocks.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+		binary.LittleEndian.PutUint64(eth1DataCountBuf, eth1Data.DepositCount)
+		eth1CountRoot := bytesutil.ToBytes32(eth1DataCountBuf)
+		fieldRoots[1] = eth1CountRoot[:]
+		if len(eth1Data.BlockHash) > 0 {
+			fieldRoots[2] = eth1Data.BlockHash
+		}
+	}
+	root, err := bitwiseMerkleize(fieldRoots, uint64(len(fieldRoots)), uint64(len(fieldRoots)))
+	if err != nil {
+		return [32]byte{}, nil
```
Comment (0xKiwi) [inline shared/stateutil/blocks.go]: same here

File: shared/stateutil/validators.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+	// Withdrawable epoch.
+	withdrawalBuf := make([]byte, 8)
+	binary.LittleEndian.PutUint64(withdrawalBuf, validator.WithdrawableEpoch)
+	withdrawalBufRoot := bytesutil.ToBytes32(withdrawalBuf)
+	fieldRoots[7] = withdrawalBufRoot[:]
+
+	root, err := bitwiseMerkleize(fieldRoots, uint64(len(fieldRoots)), uint64(len(fieldRoots)))
+	if err != nil {
+		return [32]byte{}, nil
```
Comment (0xKiwi) [inline shared/stateutil/validators.go]: same here

Comment (0xKiwi) [inline shared/stateutil/validators.go]: same here

**Comment by prestonvanloon** [comment]
I just tried this at runtime, but it didn't work in sync on an existing database.

```
[2019-12-03 17:32:41]  INFO initial-sync: Processing block 69962/161123 - estimated time remaining 506h27m0s blocksPerSecond=0.1 peers=1/1
panic: could not process blockCopy from fork choice service: could not execute state transition: could not process block: could not process block header: parent root 0x5b38ff967d056edfd5d402c1271e1a6ba48e570abb5ad82358b02b0e74e2c3e0 does not match the latest block header signing root in state 0x9553c52e55e0b7afd758e6b5e63a32dde51cf5773ef6dfe8a6ce8cb6ba1c900d

goroutine 238 [running]:
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*InitialSync).Start(0xc000171740)
        beacon-chain/sync/initial-sync/service.go:132 +0xf1c
created by github.com/prysmaticlabs/prysm/shared.(*ServiceRegistry).StartAll
        shared/service_registry.go:44 +0x23e
```

File: shared/stateutil/attestations.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+	// Target
+	targetRoot, err := checkpointRoot(data.Target)
+	if err != nil {
+		return [32]byte{}, errors.Wrap(err, "could not compute target checkpoint merkleization")
+	}
+	fieldRoots[4] = targetRoot[:]
+
+	root, err := bitwiseMerkleize(fieldRoots, uint64(len(fieldRoots)), uint64(len(fieldRoots)))
+	if err != nil {
+		return [32]byte{}, nil
```
Comment (0xKiwi) [inline shared/stateutil/attestations.go]: Can you return err here?

File: shared/stateutil/attestations.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+	return root, nil
+}
+
+func pendingAttestationRoot(att *pb.PendingAttestation) ([32]byte, error) {
+	fieldRoots := make([][]byte, 4)
+
+	// Bitfield.
+	aggregationRoot, err := bitlistRoot(att.AggregationBits, 2048)
+	if err != nil {
+		panic(err)
```
Comment (0xKiwi) [inline shared/stateutil/attestations.go]: Can you change this to return the error?

File: shared/stateutil/attestations.go [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+	}
+	fieldRoots[0] = aggregationRoot[:]
+
+	// Attestation data.
+	attDataRoot, err := attestationDataRoot(att.Data)
+	if err != nil {
+		return [32]byte{}, nil
```
Comment (0xKiwi) [inline shared/stateutil/attestations.go]: Same here

File: shared/stateutil/attestations.go [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+	// Proposer index.
+	proposerBuf := make([]byte, 8)
+	binary.LittleEndian.PutUint64(proposerBuf, att.ProposerIndex)
+	proposerRoot := bytesutil.ToBytes32(proposerBuf)
+	fieldRoots[3] = proposerRoot[:]
+
+	root, err := bitwiseMerkleize(fieldRoots, uint64(len(fieldRoots)), uint64(len(fieldRoots)))
+	if err != nil {
+		return [32]byte{}, nil
```
Comment (0xKiwi) [inline shared/stateutil/attestations.go]: Same here

File: shared/stateutil/validators.go [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+	return mixInLength(validatorsRootsRoot, validatorsRootsBufRoot), nil
+}
+
+func validatorRoot(validator *ethpb.Validator) ([32]byte, error) {
+	fieldRoots := make([][]byte, 8)
+
+	// Public key.
+	pubKeyChunks, err := pack([][]byte{validator.PublicKey})
+	if err != nil {
+		return [32]byte{}, nil
```
Comment (0xKiwi) [inline shared/stateutil/validators.go]: Same here

File: shared/stateutil/validators.go [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+	}
+	pubKeyRoot, err := bitwiseMerkleize(pubKeyChunks, uint64(len(pubKeyChunks)), uint64(len(pubKeyChunks)))
+	if err != nil {
+		return [32]byte{}, nil
```
Comment (0xKiwi) [inline shared/stateutil/validators.go]: Same here

**Pull request merged**